### PR TITLE
[Feature] cast, DIV and MOD expression JIT

### PR DESCRIPTION
**Pull request by Moonm3n**
For cast and div, mod JIT.

Related to https://github.com/StarRocks/starrocks/issues/28331, Implement arithmetic expression JIT.

## What type of PR is this:

This pr supports  JIT of multiple types of cast operations and  DIV, MOD operations.
Include: BOOLEAN, TYPE_TINYINT, TYPE_SMALLINT, TYPE_INT, TYPE_BIGINT, TYPE_LARGEINT, TYPE_FLOAT, and TYPE_DOUBLE.

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [x] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/common/status.h [hunk 1]
```diff
@@ -145,10 +145,10 @@ class STATUS_ATTRIBUTE Status {
 
     static Status RemoteFileNotFound(std::string_view msg) { return Status(TStatusCode::REMOTE_FILE_NOT_FOUND, msg); }
 
-    static Status Yield() { return {TStatusCode::YIELD, ""}; }
-
```
Comment (fzhedu) [inline be/src/common/status.h:149]: revert this change

File: be/src/exprs/cast_expr.cpp [hunk 2]
```diff
@@ -1127,6 +1134,43 @@ class VectorizedCastExpr final : public Expr {
...
+    }
+
+    StatusOr<LLVMDatum> generate_ir_impl(ExprContext* context, const llvm::Module& module, llvm::IRBuilder<>& b,
+                                         const std::vector<LLVMDatum>& datums) const override {
+        auto* l = datums[0].value;
+
+        if constexpr (FromType == TYPE_JSON || ToType == TYPE_JSON) {
+            return Status::NotSupported("JIT casting does not support JSON");
+        } else if constexpr (lt_is_decimal<FromType> || lt_is_decimal<ToType>) {
+            return Status::NotSupported("JIT casting does not support decimal");
```
Comment (fzhedu) [inline be/src/exprs/cast_expr.cpp:1149]: the `is_compilable()` can't rejet such cases?

Comment (Moonm3n) [reply be/src/exprs/cast_expr.cpp:1149]: This is for the following`std::numeric_limits<RunTimeCppType<ToType>>::max()` can be compiled.

File: be/src/exprs/arithmetic_operation.h [hunk 3]
```diff
@@ -193,58 +195,123 @@ struct ArithmeticBinaryOperator {
...
+                result.value = b.CreateLShr(l, r);
             } else {
-                return b.CreateAShr(l, r);
+                result.value = b.CreateAShr(l, r);
             }
         } else if constexpr (is_bit_shift_right_logical_op<Op>) {
-            return b.CreateLShr(l, r);
+            result.value = b.CreateLShr(l, r);
         } else {
             static_assert(is_binary_op<Op>, "Invalid binary operators");
```
Comment (fzhedu) [inline be/src/exprs/arithmetic_operation.h:311]: return an error status

File: be/src/exprs/arithmetic_operation.h [hunk 4]
```diff
@@ -193,58 +195,123 @@ struct ArithmeticBinaryOperator {
...
+                result.value = b.CreateMul(l, r);
             }
         } else if constexpr (is_div_op<Op>) {
-            // TODO(Yueyang): Support JIT compile of div operator.
-            LOG(WARNING) << "JIT compile of div operator is not supported.";
-            return nullptr;
+            llvm::Value* r_is_zero = nullptr;
+            if constexpr (lt_is_float<Type>) {
+                // return 0 when div by 0.
+                // adjusted_r = r == 0.0 ? r + 1.0 : r;
```
Comment (fzhedu) [inline be/src/exprs/arithmetic_operation.h:234]: need to consider `-0.0`?

File: be/src/exprs/arithmetic_operation.h [hunk 5]
```diff
@@ -193,58 +195,123 @@ struct ArithmeticBinaryOperator {
...
+                    auto* cond_left = b.CreateICmpEQ(l, llvm::ConstantInt::get(l->getType(), signed_minimum<LType>));
+                    auto* cond_right = b.CreateICmpEQ(r, llvm::ConstantInt::get(r->getType(), -1));
+                    auto* fpe = b.CreateAnd(cond_left, cond_right);
+                    // modify r to prevent overflow.
+                    adjusted_r = b.CreateSelect(fpe, llvm::ConstantInt::get(l->getType(), 1), adjusted_r);
+                }
+                result.value = b.CreateSDiv(l, adjusted_r);
+            }
+            result.null_flag =
+                    b.CreateSelect(r_is_zero, llvm::ConstantInt::get(result.null_flag->getType(), 1), result.null_flag);
```
Comment (fzhedu) [inline be/src/exprs/arithmetic_operation.h:259]: diff ?
```
        } else if constexpr (is_div_op<Op>) {
            // avoid 0 div a negative num, make result -0
            if constexpr (std::is_floating_point<LType>::value) {
                if (UNLIKELY(l == 0)) {
                    return 0;
                }
            }
            if constexpr (may_cause_fpe<ResultType>) {
                if (UNLIKELY(check_fpe_of_min_div_by_minus_one(l, r))) {
                    return signed_minimum<ResultType>;
                } else {
                    return l / (r + (r == 0));
                }
            } else {
                return l / (r + (r == 0));
            }
```

Comment (fzhedu) [reply be/src/exprs/arithmetic_operation.h:259]: the original code is 
```
if constexpr (may_cause_fpe<LType> && may_cause_fpe<RType>) {
```

File: be/src/exprs/arithmetic_operation.h [hunk 6]
```diff
@@ -193,58 +195,123 @@ struct ArithmeticBinaryOperator {
...
+                result.value = b.CreateSelect(r_is_negative_zero, llvm::ConstantFP::get(r->getType(), 0),
+                                              b.CreateFRem(l, adjusted_r));
+            } else {
+                // TODO(Yueyang): avoid 0 mod a negative num, make result -0
+                // adjusted_r = r == 0 ? r + 1 : r;
+                r_is_zero = b.CreateICmpEQ(r, llvm::ConstantInt::get(r->getType(), 0));
+                auto* sum = b.CreateAdd(r, llvm::ConstantInt::get(r->getType(), 1));
+                auto* adjusted_r = b.CreateSelect(r_is_zero, sum, r);
+                llvm::Value* fpe = b.getInt1(false);
+                if constexpr (may_cause_fpe<LType> && may_cause_fpe<RType>) {
```
Comment (fzhedu) [inline be/src/exprs/arithmetic_operation.h:279]: same as above

File: be/src/exprs/cast_expr.cpp [hunk 7]
```diff
@@ -1127,6 +1134,43 @@ class VectorizedCastExpr final : public Expr {
...
+                    ASSIGN_OR_RETURN(auto cast_l, IRHelper::cast_to_type(b, datum.value, ToType, FromType));
+                    llvm::Value* is_overflow = nullptr;
+                    if constexpr (lt_is_integer<FromType>) {
+                        is_overflow = b.CreateICmpNE(l, cast_l);
+                    } else if constexpr (lt_is_float<FromType>) {
+                        is_overflow = b.CreateFCmpONE(l, cast_l);
+                    }
+                    b.CreateOr(datum.null_flag, is_overflow);
+                }
+            }
```
Comment (fzhedu) [inline be/src/exprs/cast_expr.cpp:1168]: what about else?

File: be/src/exprs/arithmetic_operation.h (hunk 5 above)
Comment (Moonm3n) [reply be/src/exprs/arithmetic_operation.h:259]: Done.

File: be/src/exprs/arithmetic_operation.h (hunk 6 above)
Comment (Moonm3n) [reply be/src/exprs/arithmetic_operation.h:279]: Done.

File: be/src/exprs/cast_expr.cpp (hunk 7 above)
Comment (Moonm3n) [reply be/src/exprs/cast_expr.cpp:1168]: The else indicates that there is no overflow, no processing is required, and the two if's are to be able to compile through.

File: be/src/common/status.h (hunk 1 above)
Comment (Moonm3n) [reply be/src/common/status.h]: Done.

File: be/src/exprs/expr.cpp [hunk 8]
```diff
@@ -735,19 +735,13 @@ void Expr::get_jit_exprs(std::vector<Expr*>& exprs) {
...
-    std::vector<Expr*> next_exprs;
     if ((*expr)->should_compile()) {
         // If the current expression is compilable, we will replace it with a JITExpr.
         // This expression and its compilable subexpressions will be compiled into a single function.
         *expr = JITExpr::create(pool, *expr);
-        // Skip over the compilable subexpressions.
-        (*expr)->get_uncompilable_exprs(next_exprs);
-    } else {
-        // Search for compilable expressions in the subexpressions.
-        next_exprs = _children;
```
Comment (fzhedu) [inline be/src/exprs/expr.cpp:747]: why delete this? I don't get the point

File: be/src/exprs/arithmetic_operation.h [hunk 9]
```diff
@@ -193,58 +195,123 @@ struct ArithmeticBinaryOperator {
...
+                auto* sum = b.CreateAdd(r, llvm::ConstantInt::get(r->getType(), 1));
+                auto* adjusted_r = b.CreateSelect(r_is_zero, sum, r);
+                if constexpr (may_cause_fpe<ResultType> && may_cause_fpe<LType> && may_cause_fpe<RType>) {
+                    // fpe = l == signed_minimum<LType> && r == -1;
+                    auto* cond_left = b.CreateICmpEQ(l, llvm::ConstantInt::get(l->getType(), signed_minimum<LType>));
+                    auto* cond_right = b.CreateICmpEQ(r, llvm::ConstantInt::get(r->getType(), -1));
+                    auto* fpe = b.CreateAnd(cond_left, cond_right);
+                    // modify r to prevent overflow.
+                    adjusted_r = b.CreateSelect(fpe, llvm::ConstantInt::get(l->getType(), 1), adjusted_r);
+                }
```
Comment (fzhedu) [inline be/src/exprs/arithmetic_operation.h:255]: not equal to
```
                if (UNLIKELY(check_fpe_of_min_div_by_minus_one(l, r))) {
                    return signed_minimum<ResultType>;
```

File: be/src/exprs/arithmetic_operation.h [hunk 10]
```diff
@@ -193,58 +195,123 @@ struct ArithmeticBinaryOperator {
...
+                    auto* cond_left = b.CreateICmpEQ(l, llvm::ConstantInt::get(l->getType(), signed_minimum<LType>));
+                    auto* cond_right = b.CreateICmpEQ(r, llvm::ConstantInt::get(r->getType(), -1));
+                    fpe = b.CreateAnd(cond_left, cond_right);
+                    // modify r to prevent overflow.
+                    adjusted_r = b.CreateSelect(fpe, llvm::ConstantInt::get(l->getType(), 1), adjusted_r);
+                    result.value =
+                            b.CreateSelect(fpe, llvm::ConstantInt::get(l->getType(), 0), b.CreateSRem(l, adjusted_r));
+                } else {
+                    result.value = b.CreateSRem(l, adjusted_r);
+                }
```
Comment (fzhedu) [inline be/src/exprs/arithmetic_operation.h:290]: not equal to 
```
            if constexpr (may_cause_fpe<ResultType>) {
                if (UNLIKELY(check_fpe_of_min_div_by_minus_one(l, r))) {
                    return signed_minimum<ResultType>;
                } else {
                    return l / (r + (r == 0));
                }
```

File: be/src/exprs/cast_expr.cpp [hunk 11]
```diff
@@ -1127,6 +1134,43 @@ class VectorizedCastExpr final : public Expr {
...
+                if constexpr (std::numeric_limits<RunTimeCppType<ToType>>::max() <
+                              std::numeric_limits<RunTimeCppType<FromType>>::max()) {
+                    // Check overflow.
+                    ASSIGN_OR_RETURN(auto cast_l, IRHelper::cast_to_type(b, datum.value, ToType, FromType));
+                    llvm::Value* is_overflow = nullptr;
+                    if constexpr (lt_is_integer<FromType>) {
+                        is_overflow = b.CreateICmpNE(l, cast_l);
+                    } else if constexpr (lt_is_float<FromType>) {
+                        is_overflow = b.CreateFCmpONE(l, cast_l);
+                    }
```
Comment (fzhedu) [inline be/src/exprs/cast_expr.cpp:1165]: what about downward overflowing?

File: be/src/exprs/expr.cpp (hunk 8 above)
Comment (Moonm3n) [reply be/src/exprs/expr.cpp:747]: If this section of code is not deleted, when replacing ﻿expr, if the replaced node is not at the top level, it cannot truly affect the ﻿expr tree. This issue was introduced when supporting more complex compilable ﻿expr tree rules.

File: be/src/exprs/arithmetic_operation.h (hunk 9 above)
Comment (Moonm3n) [reply be/src/exprs/arithmetic_operation.h:255]: The code here is not exactly equivalent, but it should yield the same result. The functionality of the original ﻿if statement is implemented by modifying the divisor. This is done because constructing an ﻿if statement here is difficult, and in a ternary expression, both expressions will be executed, potentially causing a division by zero error. Therefore, the divisor is modified prior to the division operation to ensure correctness.

File: be/src/exprs/arithmetic_operation.h (hunk 10 above)
Comment (Moonm3n) [reply be/src/exprs/arithmetic_operation.h:290]: The code here is not exactly equivalent, but it should yield the same result. The functionality of the original ﻿if statement is implemented by modifying the divisor. This is done because constructing an ﻿if statement here is difficult, and in a ternary expression, both expressions will be executed, potentially causing a division by zero error. Therefore, the divisor is modified prior to the division operation to ensure correctness

File: be/src/exprs/cast_expr.cpp (hunk 11 above)
Comment (Moonm3n) [reply be/src/exprs/cast_expr.cpp:1165]: Is there an issue with this 'if' check? `if constexpr ((std::numeric_limits<RunTimeCppType<ToType>>::max() <`

File: be/src/exprs/expr.cpp [hunk 12]
```diff
@@ -213,6 +213,33 @@ Status Expr::create_expr_tree(ObjectPool* pool, const TExpr& texpr, ExprContext*
...
+                   << apache::thrift::ThriftDebugString(texpr);
+        // Fall back to the non-JIT path.
+        return Status::OK();
+    }
+
+    if (e != prev_e) {
+        // The root node was replaced, so we need to update the context.
+        *ctx = pool->add(new ExprContext(e));
+    }
+
```
Comment (fzhedu) [inline be/src/exprs/expr.cpp:242]: delete this, as done in `create_tree_from_thrift_with_jit`

File: be/src/exprs/cast_expr.cpp [hunk 13]
```diff
@@ -1127,6 +1134,45 @@ class VectorizedCastExpr final : public Expr {
...
+                               std::numeric_limits<RunTimeCppType<FromType>>::max()) ||
+                              (std::numeric_limits<RunTimeCppType<ToType>>::min() >
+                               std::numeric_limits<RunTimeCppType<FromType>>::min())) {
+                    // Check overflow.
+                    ASSIGN_OR_RETURN(auto cast_l, IRHelper::cast_to_type(b, datum.value, ToType, FromType));
+                    llvm::Value* is_overflow = nullptr;
+                    if constexpr (lt_is_integer<FromType>) {
+                        is_overflow = b.CreateICmpNE(l, cast_l);
+                    } else if constexpr (lt_is_float<FromType>) {
+                        is_overflow = b.CreateFCmpONE(l, cast_l);
```
Comment (fzhedu) [inline be/src/exprs/cast_expr.cpp:1166]: a doube is cast to float, then cast to double, this equal may be not true?

Comment (fzhedu) [reply be/src/exprs/cast_expr.cpp:1166]: a float is cast to integers, then cast to float, the equal also my be not true

File: be/src/exprs/cast_expr.cpp (hunk 11 above)
Comment (Moonm3n) [reply be/src/exprs/cast_expr.cpp:1167]: Done.

File: be/src/exprs/expr.cpp (hunk 12 above)
Comment (Moonm3n) [reply be/src/exprs/expr.cpp:242]: Done.

File: be/src/testutil/exprs_test_helper.h [hunk 14]
```diff
@@ -188,10 +189,43 @@ class ExprsTestHelper {
...
+
+        Chunk chunk;
+        chunk.append_column(ptr, 0);
+        auto jit_ptr = jit_expr->evaluate(&exprContext, &chunk);
+
+        ASSERT_TRUE(jit_ptr->is_constant() == ptr->is_constant());
+
+        // ASSERT_TRUE(jit_ptr->is_nullable() == ptr->is_nullable());
+        ASSERT_TRUE(jit_ptr->size() == ptr->size());
+        jit_ptr->equals(0, *ptr, jit_ptr->size());
```
Comment (fzhedu) [inline be/src/testutil/exprs_test_helper.h:226]: should assert_true

File: be/src/exprs/jit/jit_engine.cpp [hunk 15]
```diff
@@ -223,9 +219,13 @@ void JITEngine::print_module(const llvm::Module& module) {
...
 }
 
-void* JITEngine::lookup_function(const std::string& expr_name) {
+void* JITEngine::lookup_function(const std::string& expr_name, bool mast_exist) {
     auto addr = _jit->lookup(expr_name);
     if (UNLIKELY(!addr || UNLIKELY(addr->isNull()))) {
+        if (!mast_exist) {
+            return nullptr;
+        }
+
```
Comment (fzhedu) [inline be/src/exprs/jit/jit_engine.cpp:228]: change LOG(ERROR) at Line:233 to ROW_LOG << xxx

File: be/src/exprs/jit/jit_engine.cpp [hunk 16]
```diff
@@ -223,9 +219,13 @@ void JITEngine::print_module(const llvm::Module& module) {
     LOG(INFO) << "JIT: Generated IR:\n" << str;
 }
 
-void* JITEngine::lookup_function(const std::string& expr_name) {
+void* JITEngine::lookup_function(const std::string& expr_name, bool mast_exist) {
```
Comment (fzhedu) [inline be/src/exprs/jit/jit_engine.cpp:222]: must_exist?

File: be/src/exprs/jit/ir_helper.h [hunk 17]
```diff
@@ -30,7 +30,7 @@ struct LLVMDatum {
     llvm::Value* value = nullptr;     ///< Represents the actual value of the datum.
     llvm::Value* null_flag = nullptr; ///< Represents the nullity status of the datum.
 
-    LLVMDatum(llvm::IRBuilder<>& b) { null_flag = llvm::ConstantInt::get(b.getInt8Ty(), 0); }
+    LLVMDatum(llvm::IRBuilder<>& b) { null_flag = llvm::ConstantInt::get(b.getInt8Ty(), 0, false); }
```
Comment (fzhedu) [inline be/src/exprs/jit/ir_helper.h:33]: why add `false`?

Comment (Moonm3n) [reply be/src/exprs/jit/ir_helper.h:33]: In LLVM, all types are unsigned, making an explicit declaration here. This is primarily to differentiate from signed numbers, as when creating an LLVM ConstantInt using literals, a signed 0 will be misconstrued as an unsigned 0, causing an error.

File: be/src/exprs/jit/jit_engine.cpp (hunk 16 above)
Comment (Moonm3n) [reply be/src/exprs/jit/jit_engine.cpp:222]: The main purpose here is to reduce the output of logs.

File: be/src/exprs/arithmetic_expr.cpp [hunk 18]
```diff
@@ -191,11 +184,28 @@ class VectorizedDivArithmeticExpr final : public Expr {
         }
     }
 
-    bool is_compilable() const override { return false; }
+    bool is_compilable() const override { return Type != TYPE_LARGEINT && IRHelper::support_jit(Type); }
```
Comment (fzhedu) [inline be/src/exprs/arithmetic_expr.cpp:187]: all arithmetic op disable largeint?

Comment (Moonm3n) [reply be/src/exprs/arithmetic_expr.cpp:187]: Only the div operator is disabled here.

File: be/src/exprs/jit/jit_engine.cpp [hunk 19]
```diff
@@ -93,26 +85,30 @@ StatusOr<JITScalarFunction> JITEngine::compile_scalar_function(ExprContext* cont
...
+
     // Generate scalar function IR.
     RETURN_IF_ERROR(JITFunction::generate_scalar_function_ir(context, *module, expr));
-    if (llvm::verifyModule(*module)) {
-        return Status::JitCompileError("Failed to generate scalar function IR");
+    std::string error;
+    llvm::raw_string_ostream errs(error);
+    if (llvm::verifyModule(*module, &errs)) {
+        return Status::JitCompileError(
+                fmt::format("Failed to generate scalar function IR, verify failed: {}", errs.str()));
```
Comment (fzhedu) [inline be/src/exprs/jit/jit_engine.cpp:104]: Failed to generate scalar function IR, verify failed =>
Failed to generate scalar function IR, errors:

File: be/src/exprs/jit/jit_engine.cpp [hunk 20]
```diff
@@ -93,26 +85,30 @@ StatusOr<JITScalarFunction> JITEngine::compile_scalar_function(ExprContext* cont
...
     }
 
     // Optimize module.
     instance->optimize_module(module.get());
-    if (llvm::verifyModule(*module)) {
-        return Status::JitCompileError("Failed to optimize scalar function IR");
+    if (llvm::verifyModule(*module, &errs)) {
+        return Status::JitCompileError(
+                fmt::format("Failed to optimize scalar function IR, verify failed: {}", errs.str()));
```
Comment (fzhedu) [inline be/src/exprs/jit/jit_engine.cpp:111]: diit

File: be/src/exprs/jit/jit_engine.cpp (hunk 16 above)
Comment (fzhedu) [reply be/src/exprs/jit/jit_engine.cpp:222]: mask_exist or must ? seems a typo

File: be/test/exprs/cast_expr_test.cpp [hunk 21]
```diff
@@ -276,31 +287,131 @@ TEST_F(VectorizedCastExprTest, stringLiteralIntCastToBoolean) {
...
+        ExprsTestHelper::verify_result_with_jit(ptr, expr.get(), runtime_state);
+    }
+}
+
+TEST_F(VectorizedCastExprTest, numericJITCast) {
+    numeric_cast_with_jit<TYPE_TINYINT, TYPE_TINYINT>(&runtime_state, expr_node);
+    numeric_cast_with_jit<TYPE_TINYINT, TYPE_SMALLINT>(&runtime_state, expr_node);
+    numeric_cast_with_jit<TYPE_TINYINT, TYPE_INT>(&runtime_state, expr_node);
+    numeric_cast_with_jit<TYPE_TINYINT, TYPE_BIGINT>(&runtime_state, expr_node);
+    // numeric_cast_with_jit<TYPE_TINYINT, TYPE_LARGEINT>(&runtime_state, expr_node);
```
Comment (fzhedu) [inline be/test/exprs/cast_expr_test.cpp:334]: delete them

Comment (Moonm3n) [reply be/test/exprs/cast_expr_test.cpp]: Fixed.

File: be/src/exprs/jit/jit_engine.cpp (hunk 20 above)
Comment (Moonm3n) [reply be/src/exprs/jit/jit_engine.cpp]: Fixed.

File: be/src/exprs/jit/jit_engine.cpp (hunk 19 above)
Comment (Moonm3n) [reply be/src/exprs/jit/jit_engine.cpp]: Fixed.

File: be/src/exprs/jit/jit_engine.cpp (hunk 15 above)
Comment (Moonm3n) [reply be/src/exprs/jit/jit_engine.cpp:226]: Fixed.

File: be/src/exprs/cast_expr.cpp (hunk 13 above)
Comment (Moonm3n) [reply be/src/exprs/cast_expr.cpp]: Fixed.

**Review by fzhedu — DISMISSED** [review]

File: be/src/testutil/exprs_test_helper.h [hunk 22]
```diff
@@ -172,26 +173,66 @@ class ExprsTestHelper {
...
+        Chunk chunk;
+        chunk.append_column(ptr, 0);
+        ptr = jit_expr->evaluate(&exprContext, &chunk);
         // Verify the result after JIT.
         test_func(ptr);
+
+        Expr::close(expr_ctxs, runtime_state);
+    }
+
+    static void verify_result_with_jit(const ColumnPtr& ptr, Expr* expr, RuntimeState* runtime_state) {
```
Comment (kevincai) [inline be/src/testutil/exprs_test_helper.h:207]: not sure if this is ok to put into be/src/testutil/ with asserts and checks.

it may be better moved into be/test directory into a new header file. testutil in be/src/ should be modified as little as possible whenever the added code can be moved to be/test.

Comment (fzhedu) [reply be/src/testutil/exprs_test_helper.h:207]: better move to be/test/xxx

**Review by kevincai — APPROVED** [review]

**Review by fzhedu — APPROVED** [review]

**Review by satanson — APPROVED** [review]
